### PR TITLE
Rework Debug CodeLens to display on correct documents

### LINKED ISSUE #178: CodeLens "Debug" appears on wrong document
### Environment

1. `node -v`: v8.9.1
2. `npm -v`: 5.5.1
4. `npm ls react-scripts` (if you haven’t ejected): N/A
1. Operating system: Windows 7 x64
1. VS Code: 1.18.1
1. Jest extension: 4.5.0

### Steps to Reproduce

1. Clone the repo:
    ```
    git clone https://github.com/seanpoulter/vscode-jest-178
    ```
1. Open the repo in Visual Studio Code
1. Open `index.js` and `index.test.js`
1. Wait for the CodeLens "Debug" to appear in `index.test.js`
1. Switch to another window

### Expected Behavior

The "Debug" should **only** appear on the appropriate document.

### Actual Behavior

"Debug" appears on the wrong document. Here's a GIF of it appearing correctly on `index.test.js` then incorrectly on `index.js`.

2017-12-01 -- vscode-jest 178 - debug appears in wrong window

### PR DESCRIPTION
**Pull request by seanpoulter**
The Debug CodeLensProvider incorrectly adds CodeLenses with decorations that may or may not have been updated for the active text editor, and does not handle the case when there is more than one visible text editor.

This change shuffles around the Debug CodeLens files, and moves the responsibility for fetching and caching test results into a new `TestResultProvider` class. The relevant methods to parse, join, and sort
test results have been extracted from `JestExt`.

To improve support for multiple visible text editors, when a test run is complete `updateWithData` will call `triggerUpdateDecorations` for all visible text editors.

This fixes #178.
<br/>

No rush on this one, it's a meaty change. I'd rather we make sure we're invalidating the test result cache is invalidated properly before we ship.

### CODE REVIEW

**Comment by seanpoulter** [comment]
@orta, the changelog check is working. :relaxed:

**Comment by orta** [comment]
Ace 👍

File: src/JestExt.ts [hunk 1]
```diff
@@ -599,4 +527,22 @@ export class JestExt {
...
+  }
+
+  removeCachedTestResults(document: vscode.TextDocument) {
+    if (!document || document.isUntitled) {
+      return
+    }
+
+    const filePath = document.fileName
+    this.testResultProvider.removeCachedResults(filePath)
+  }
```
Comment (orta) [inline src/JestExt.ts]: nice, breaking up the monolith

Comment (seanpoulter) [reply src/JestExt.ts]: Thanks. Figured I'd practice some test-driven development. I'm happy we both like how it forces you to add seams to write tests. I'm curious if they'll help integrate changes in the long run or just end up "too fragile".

Comment (orta) [reply src/JestExt.ts]: likely the former, people can trust their changes work as expected

File: src/extension.ts [hunk 2]
```diff
@@ -47,13 +47,15 @@ export function activate(context: vscode.ExtensionContext) {
...
+    vscode.languages.registerCodeLensProvider(languages, extensionInstance.debugCodeLensProvider),
     vscode.workspace.onDidChangeConfiguration(e => {
       if (e.affectsConfiguration('jest')) {
         const updatedSettings = getExtensionSettings()
         extensionInstance.triggerUpdateSettings(updatedSettings)
       }
-    })
+    }),
+
+    vscode.workspace.onDidCloseTextDocument(extensionInstance.onDidCloseTextDocument)
```
Comment (seanpoulter) [inline src/extension.ts]: This is throwing an error from `onDidCloseTextDocument()` because `this` is `null`. :angry:

**Comment by orta** [comment]
Alright, things look good to me. I'll clone it and have a test look on some projects before shipping - nice work

**Comment by seanpoulter** [comment]
Thanks for reviewing all that. Let me know if you find anything, and if you've got any preferences on the next issues to tackle. I'm trying to get #116 and #177 in progress.

**Comment by orta** [comment]
looks good - shipped. #116 looks important

**Pull request merged**